frontend/src/features/admin-form/create/workflow/components/WorkflowContent/EditStepBlock/RespondentBlock.tsx

```diff
@@ -170,19 +204,426 @@ const DynamicRespondentOption = ({
+    selectedConditionalFieldOptions: string[],
+  ) => {
+    if (
+      checkIsOptionsMismatched(
+        optionsToRecipientsMapOptions,
+        selectedConditionalFieldOptions,
+      )
+    ) {
+      return CONDITIONAL_ROUTING_MISMATCHED_OPTIONS_ERROR_MESSAGE
+    }
+  }
+
+  const validateOptionsToRecipientsMapErrorMessage = validateOptions(
+    [...Object.keys(selectedConditionalFieldOptionsToRecipientsMap || {})],
+    selectedConditionalField?.fieldOptions || [],
+  )
+
+  const noEmailToOptionsMappingErrorMessage =
+    !selectedConditionalFieldOptionsToRecipientsMap
+      ? 'You must add email(s) to options before saving this step'
+      : null
+
+  const validateCsvFile = async (
+    file: File | null,
+  ): Promise<string | undefined> => {
+    if (!file) return 'Please upload a CSV file'
+
+    let conditionalRoutingCsvString
+    try {
+      conditionalRoutingCsvString = await parseCsvFileToCsvString(
+        file,
+        (headerRow) => {
+          return {
+            isValid:
+              headerRow &&
+              headerRow.length === 2 &&
+              headerRow[0] === 'Options' &&
+              headerRow[1] === 'Add email(s) in this column',
+            invalidReason:
+              'Your CSV file should only contain 2 columns with the "Options" and "Add email(s) in this column" headers',
+          }
+        },
+      )
+    } catch (error) {
+      return (error as Error).message
+    }
+
+    const options = conditionalRoutingCsvString.split('\r\n')
+    const optionsSet = new Set<string>()
+
+    for (const row of options) {
+      const [option, ...recipients] = row.split(',')
+      if (recipients.length <= 0 || !recipients[0] || !option) {
+        return CONDITIONAL_ROUTING_EMAILS_OPTIONS_MISSING_ERROR_MESSAGE
+      }
+      if (recipients.some((recipient) => !isEmail(recipient))) {
+        return CONDITIONAL_ROUTING_INVALID_CSV_FORMAT_ERROR_MESSAGE
+      }
+      optionsSet.add(option)
+    }
+
+    const selectedConditionalFieldOptions =
+      selectedConditionalField?.fieldOptions
+
+    if (optionsSet.size < options.length) {
+      return CONDITIONAL_ROUTING_DUPLICATE_OPTIONS_ERROR_MESSAGE
+    }
+
+    return validateOptions(
+      [...optionsSet],
+      selectedConditionalFieldOptions || [],
+    )
+  }
+
+  return (
+    <>
+      <ConditionalRoutingMappingDeleteModal
+        isOpen={isDeleteConfirmModalOpen}
+        onClose={() => setIsDeleteConfirmModalOpen(false)}
+        handleDelete={removeOptionsToRecipientsMapping}
+      />
+      <ConditionalRoutingOptionModal
+        conditionalFieldItems={conditionalFieldItems}
+        isLoading={isLoading}
+        isOpen={isOpen}
+        onClose={onClose}
+        control={conditionalRoutingConfigControl}
+        errors={conditionalRoutingConfigErrors}
+        onDownloadCsvClick={handleSkeletonCsvDownload(formId)}
+        onSubmit={handleSubmit(
+          handleConditionalRoutingConfigSubmit(watch('conditional_field')),
```

Comment:
Might be able to utilise `getValues('conditional_field')` in `handleConditionalRoutingConfigSubmit()` instead to reduce opportunities of inefficiencies. -- I generally try to avoid using `watch` if possible as it is sometimes too powerful and gets misused.

Reply:
fixed in [f90db28](https://github.com/opengovsg/FormSG/pull/7804/commits/f90db2817e67d36d2a4b35f4add76ba25885d5db)